lib/cretonne/src/ssa.rs

```diff
@@ -0,0 +1,716 @@
+//! A SSA-building API that handles incomplete CFGs.
+//!
+//! The algorithm is based upon Braun M., Buchwald S., Hack S., Leißa R., Mallon C.,
+//! Zwinkau A. (2013) Simple and Efficient Construction of Static Single Assignment Form.
+//! In: Jhala R., De Bosschere K. (eds) Compiler Construction. CC 2013.
+//! Lecture Notes in Computer Science, vol 7791. Springer, Berlin, Heidelberg
+
+use ir::{Ebb, Value, Inst, Type, DataFlowGraph};
+use std::hash::Hash;
+use entity_map::{EntityMap, EntityRef, PrimaryEntityData};
+use sparse_map::{SparseMap, SparseMapValue};
+use packed_option::ReservedValue;
+use std::u32;
+use std::collections::HashMap;
+
+/// Structure containing the data relevant the construction of SSA for a given function.
+///
+/// The parameter struct `Variable` corresponds to the way variables are represented in the
+/// non-SSA language you're translating from.
+///
+/// The SSA building relies on information about the variables used and defined, as well as
+/// their position relative to basic blocks which are stricter than extended basic blocks since
+/// they don't allow branching in the middle of them.
+///
+/// This SSA building module allows you to def and use variables on the fly while you are
+/// constructing the CFG, no need for a separate SSA pass after the CFG is completed.
+///
+/// A basic block is said _filled_ if all the instruction that it contains have been translated,
+/// and it is said _sealed_ if all of its predecessors have been declared. Only filled predecessors
+/// can be declared.
+pub struct SSABuilder<Variable>
+    where Variable: EntityRef
+{
+    // Records for every variable its type and, for every revelant block, the last definition of
+    // the variable in the block.
+    variables: SparseMap<Variable, VariableData<Variable>>,
+    // Records the position of the basic blocks and the list of values used but not defined in the
+    // block.
+    blocks: EntityMap<Block, BlockData<Variable>>,
+    // Records the basic blocks at the beginning of the `Ebb`s.
+    // The value is `(Ebb,Block)` because in a `SparseMap`, the key has to be computable from the
+    // value.
+    ebb_headers: SparseMap<Ebb, (Ebb, Block)>,
+}
+
+// Describes the current position of a basic block in the control flow graph.
+enum BlockData<Variable> {
+    // A block at the top of an `Ebb`. Contains the list of known predecessors of this block
+    // and a boolean that indicates if the block is sealed or not.
+    // A block is sealed if all of its predecessors have been declared.
+    // The fields are `(predecessors, sealed, ebb, undef_values)`.
+    EbbHeader(HashMap<Block, Inst>, bool, Ebb, Vec<(Variable, Value)>),
```

Comment:
This is a little too complex for a tuple to my taste. I would prefer a struct with named fields.